components/org.wso2.carbon.database.utils/src/main/java/org/wso2/carbon/database/utils/jdbc/Template.java

```diff
@@ -0,0 +1,262 @@
+                        JdbcUtils.getConvertedAutoGeneratedColumnName(dbProductName, ID)})) {
+                    doInternalUpdate(queryFilter, preparedStatement);
+                    logDebug("Mapping generated key (Auto Increment ID) to the object");
+                    try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
+                        if (generatedKeys.next()) {
+                            resultId = generatedKeys.getInt(1);
+                            logDebug("Newly inserted ID (Auto Increment ID) is {} for the bean {} ",
+                                    resultId, bean);
+                        } else {
+                            throw new SQLException(JdbcConstants.ErrorCodes.ERROR_CODE_AUTO_GENERATED_ID_FAILURE.
+                                    getErrorMessage());
+                        }
+                    }
+                }
+                return resultId;
+            } else {
+                try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+                    doInternalUpdate(queryFilter, preparedStatement);
+                }
+            }
+        } catch (SQLException e) {
+            logDebug("Error in performing database insert: {} with parameters {}", query, queryFilter);
+            throw new DataAccessException(JdbcConstants.ErrorCodes.ERROR_CODE_DATABASE_QUERY_PERFORMING.
+                    getErrorMessage() + query, JdbcConstants.ErrorCodes.ERROR_CODE_DATABASE_QUERY_PERFORMING.
+                    getErrorCode(), e);
+        }
+        return 0;
+    }
+
+    /**
+     * Executes the jdbc insert/update query.
+     *
+     * @param query       The SQL for insert/update.
+     * @param queryFilter Query filter to prepared statement parameter binding.
+     * @param bean        the Domain object to be inserted/updated.
+     */
+    public <T extends Object> int executeBatchInsert(String query, QueryFilter queryFilter, T bean)
+            throws DataAccessException {
+
+        Connection connection = TransactionManager.getTransactionEntry().getConnection();
+        try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            doInternalBatchUpdate(queryFilter, preparedStatement);
+        } catch (SQLException e) {
+            logDebug("Error in performing database insert: {} with parameters {}", query, queryFilter);
+            throw new DataAccessException(JdbcConstants.ErrorCodes.ERROR_CODE_DATABASE_QUERY_PERFORMING_ERROR
+                    .getErrorMessage() + query, e);
+        }
+        return 0;
+    }
+
+    /**
+     * Executes a query on JDBC and return the result as a list of domain objects.
+     *
+     * @param query     the SQL query with the parameter placeholders.
+     * @param rowMapper Row mapper functional interface.
+     * @return List of domain objects of required type.
+     * @see #executeQuery(String, RowMapper, QueryFilter)
```

Comment:
Shall we complete the method comment with throws description and fix other places?